src/providers/FilterProvider.tsx

```diff
@@ -0,0 +1,182 @@
+// SPDX-FileCopyrightText: 2024 PNED G.I.E.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+"use client";
+
+import { ActiveFilter, Filter } from "@/services/discovery/types/filter.type";
+import React, { createContext, useContext, useEffect, useReducer } from "react";
+
+type FilterContextState = {
+  filters: Filter[];
+  activeFilters: ActiveFilter[];
+  isLoading: boolean;
+  error: { message: string; statusCode: number } | null;
+};
+
+type FilterContextReturnType = FilterContextState & {
+  addActiveFilter: (filter: ActiveFilter) => void;
+  removeActiveFilter: (key: string, source: string) => void;
+  clearActiveFilters: () => void;
+};
+
+enum FilterActionType {
+  FILTERS_RETRIEVED,
+  ACTIVE_FILTER_ADDED,
+  ACTIVE_FILTER_REMOVED,
+  ACTIVE_FILTERS_CLEARED,
+  ACTIVE_VALUE_REMOVED,
+  LOADING,
+  REJECTED,
+}
+
+type FilterAction = {
+  type: FilterActionType;
+  payload?: unknown;
+};
+
+function reducer(
+  state: FilterContextState,
+  action: FilterAction
+): FilterContextState {
+  switch (action.type) {
+    case FilterActionType.FILTERS_RETRIEVED: {
```

Comment:
Curious is there any specific reason to use redux pattern here?

All i see is that you have a list of active filters and you need to add or remove values from it. 

Isnt this pattern makes it complex?

You should be able to directly update the states from your callbacks if you keep them as individual states

  const [filters, setFilters] = useState([]);
  const [activeFilters, setActiveFilters] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(null);

What are your thoughts on this?

Reply:
I don't have strong opinion on that. You are right, I can use multiple states as you proposed.

I personally like using Provider + Reducer pattern. The reducer allows to have all the logic in a single function and we can batch state changes so that we reduce the number of re-renders.

Reply:
I also like reducer pattern and i believe its a good technique for handling very complex states. 

However FilterProvider for me is not that complex. So IMO its not worth it

Reply:
For the complexity, I don't think it is increased here. I personally find that it makes it clear what are the different state transitions that can happen. The logic itself should stay the same, just be scattered inside the different event handlers

Reply:
I agree that it's currently not very complex. But it may get a little bit more complex with the upcoming `variant` filter

Reply:
if you strongly feel to keep it, you can keep this logic. (if you forsee more complexity coming in)

I've used reducers for managing the whole application state centrally in the past. Using it for individual feature was a bit new to me

Reply:
Okay thanks, let's keep it then